File: ios/Planty/Models/Plant.swift
import Foundation

/// One growing thing in one place. Mirrors internal/plant.Plant on the wire.
struct Plant: Codable, Sendable, Hashable, Identifiable {
    let id: UUID
    let slug: String

    var commonName: String
    var botanicalName: String?
    var variety: String?

    var domain: PlantDomain
    var steward: String
    var status: PlantStatus

    var location: String
    var haArea: String?

    var accessibility: PlantAccessibility
    var wateringMethod: WateringMethod
    var letpotDripper: Int?

    var potSizeIn: Double?
    var potMaterial: String?
    var hasDrainage: Bool?
    var soilMix: String?
    var lightExposure: LightExposure?

    var minTempF: Double?
    var careProfile: CareProfile

    var acquiredAt: Date?
    var archivedAt: Date?

    /// Set while it is indoors for the cold. Where it is right now beats
    /// `location`, which is the place it goes back to.
    var shelteredAt: Date?
    var createdAt: Date
    var updatedAt: Date

    /// The newest photograph of this plant, short-lived and only sent by the
    /// list. Optional because a plant nobody has photographed yet is normal,
    /// and because every other endpoint returns a plant without one.
    var photoURL: URL?
    var photoTakenAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case slug
        case commonName = "common_name"
        case botanicalName = "botanical_name"
        case variety
        case domain
        case steward
        case status
        case location
        case haArea = "ha_area"
        case accessibility
        case wateringMethod = "watering_method"
        case letpotDripper = "letpot_dripper"
        case potSizeIn = "pot_size_in"
        case potMaterial = "pot_material"
        case hasDrainage = "has_drainage"
        case soilMix = "soil_mix"
        case lightExposure = "light_exposure"
        case minTempF = "min_temp_f"
        case careProfile = "care_profile"
        case acquiredAt = "acquired_at"
        case archivedAt = "archived_at"
        case shelteredAt = "sheltered_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case photoURL = "photo_url"
        case photoTakenAt = "photo_taken_at"
    }
}

extension Plant {
    static let stewardSelf = "self"

    /// Somebody else owns this. Drives the purple badge and the sort, and
    /// nothing else: ownership is metadata, never severity.
    var isFriends: Bool { !steward.isEmpty && steward != Self.stewardSelf }

    var ownerName: String? { isFriends ? steward : nil }

    /// Indoors for the cold right now, which is worth saying because the cold
    /// warning repeats until somebody records that it was answered.
    var isSheltered: Bool { shelteredAt != nil }

    /// Only a plant with a cold threshold was ever asked to come in, so only
    /// one of those can sensibly be marked as having gone back out.
    var canShelter: Bool { minTempF != nil && archivedAt == nil }

    /// Friend plants read as "Maya's plant"; mine only says so when ambiguous.
    var ownershipLabel: String {
        isFriends ? "\(steward)'s plant" : "Mine"
    }

    var ownershipAccessibilityLabel: String {
        isFriends ? "Friend's plant, belongs to \(steward)" : "My plant"
    }

    /// A due action here depends on a human remembering it happened.
    var needsChasing: Bool { wateringMethod == .hand }

    /// Likelihood of neglect, not current thirst. Mirrors Plant.Risk() in Go so
    /// the app orders a list the same way the service does.
    var risk: Int {
        var risk = 0
        if needsChasing { risk += 2 }
        switch accessibility {
        case .hard: risk += 2
        case .awkward: risk += 1
        default: break
        }
        if isFriends { risk += 2 }
        if status == .struggling { risk += 1 }
        return risk
    }

    var displaySpecies: String? {
        if let variety, let botanicalName {
            return "\(botanicalName) '\(variety)'"
        }
        return botanicalName ?? variety
    }
}

/// Descriptive care knowledge. Every field optional: most are unknown when a
/// plant is first written down, and a prompt reads this, not a WHERE clause.
struct CareProfile: Codable, Sendable, Hashable {
    var notes: String?
    var ownerSays: String?
    var quirks: String?
    var wantsLight: LightExposure?

    var moistureLow: Double?
    var moistureHigh: Double?
    var humidityPref: String?
    var dormancy: String?

    var daysToMaturity: Int?
    var sownAt: Date?
    var transplantedAt: Date?
    var feedSchedule: String?
    var expectedHarvest: Date?
    var needsPollination: Bool?

    var hardinessZone: String?
    var frostSensitive: Bool?
    var successionDays: Int?
    var lastFrostExpected: Date?
    var firstFrostExpected: Date?

    enum CodingKeys: String, CodingKey {
        case notes
        case ownerSays = "owner_says"
        case quirks
        case wantsLight = "wants_light"
        case moistureLow = "moisture_low"
        case moistureHigh = "moisture_high"
        case humidityPref = "humidity_pref"
        case dormancy
        case daysToMaturity = "days_to_maturity"
        case sownAt = "sown_at"
        case transplantedAt = "transplanted_at"
        case feedSchedule = "feed_schedule"
        case expectedHarvest = "expected_harvest"
        case needsPollination = "needs_pollination"
        case hardinessZone = "hardiness_zone"
        case frostSensitive = "frost_sensitive"
        case successionDays = "succession_days"
        case lastFrostExpected = "last_frost_expected"
        case firstFrostExpected = "first_frost_expected"
    }
}

extension Plant {
    /// Only the list endpoint sends a photo URL, so a copy that arrived from
    /// any other call must not blank the picture already on screen.
    func keepingPhoto(from previous: Plant) -> Plant {
        guard photoURL == nil else { return self }
        var kept = self
        kept.photoURL = previous.photoURL
        kept.photoTakenAt = previous.photoTakenAt
        return kept
    }
}

File: internal/plant/plant.go
// Package plant holds the domain types shared by the store, the HTTP API, the
// Dusk plugin and the iOS app. The reasoning behind the shape is in
// docs/DATA-MODEL.md.
package plant

import (
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Domain decides which questions are worth asking: a houseplant is judged on
// staying alive, a tomato on bearing fruit.
type Domain string

const (
	DomainHouseplant    Domain = "houseplant"
	DomainEdibleIndoor  Domain = "edible_indoor"
	DomainEdibleOutdoor Domain = "edible_outdoor"
)

// Status marks dead as a state rather than a deletion, because what was done to
// a plant that died is the most useful record for not repeating it.
type Status string

const (
	StatusAlive      Status = "alive"
	StatusStruggling Status = "struggling"
	StatusDormant    Status = "dormant"
	StatusDead       Status = "dead"
	StatusGone       Status = "gone"
)

// ValidateArchive rejects states that contradict an archived record.
func (s Status) ValidateArchive() error {
	if s != StatusDead && s != StatusGone {
		return invalid("archiving records dead or gone, not %q", s)
	}
	return nil
}

// WateringMethod decides whether Planty can act, or only report.
type WateringMethod string

const (
	WateringLetPot WateringMethod = "letpot"
	WateringHand   WateringMethod = "hand"
)

// Accessibility drives sensor priority; unreachable plants stop being checked.
type Accessibility string

const (
	AccessEasy    Accessibility = "easy"
	AccessAwkward Accessibility = "awkward"
	AccessHard    Accessibility = "hard"
)

// LightExposure is what the plant receives. What it wants lives in CareProfile.
type LightExposure string

const (
	LightDirect         LightExposure = "direct"
	LightBrightIndirect LightExposure = "bright_indirect"
	LightMedium         LightExposure = "medium"
	LightLow            LightExposure = "low"
)

// StewardSelf marks a plant as the operator's own; anything else is a name.
const StewardSelf = "self"

// ErrInvalid marks a failure the caller caused, so a handler can answer 400
// without every call site guessing. A server fault must never report as one.
var ErrInvalid = errors.New("invalid")

// invalid wraps a rejection so it carries whose fault it is.
func invalid(format string, args ...any) error {
	return fmt.Errorf("%w: %s", ErrInvalid, fmt.Sprintf(format, args...))
}

// Plant is one growing thing in one place. Fields automations query are typed;
// descriptive care knowledge goes in CareProfile.
type Plant struct {
	ID   uuid.UUID `json:"id"`
	Slug string    `json:"slug"`

	CommonName    string `json:"common_name"`
	BotanicalName string `json:"botanical_name,omitempty"`
	Variety       string `json:"variety,omitempty"`

	Domain  Domain `json:"domain"`
	Steward string `json:"steward"`
	Status  Status `json:"status"`

	Location string `json:"location"`
	HAArea   string `json:"ha_area,omitempty"`

	Accessibility  Accessibility  `json:"accessibility"`
	WateringMethod WateringMethod `json:"watering_method"`
	LetPotDripper  *int           `json:"letpot_dripper,omitempty"`

	PotSizeIn     *float64      `json:"pot_size_in,omitempty"`
	PotMaterial   string        `json:"pot_material,omitempty"`
	HasDrainage   *bool         `json:"has_drainage,omitempty"`
	SoilMix       string        `json:"soil_mix,omitempty"`
	LightExposure LightExposure `json:"light_exposure,omitempty"`

	// Typed and indexed rather than left in CareProfile because the cold-snap
	// automation queries it nightly, where a JSON key typo would fail silently.
	MinTempF *float64 `json:"min_temp_f,omitempty"`

	CareProfile CareProfile `json:"care_profile"`
	Toxicity    Toxicity    `json:"toxicity"`

	AcquiredAt *time.Time `json:"acquired_at,omitempty"`
	ArchivedAt *time.Time `json:"archived_at,omitempty"`

	// Set when carried indoors ahead of a cold night, so the system can also
	// say when to put it back out.
	ShelteredAt *time.Time `json:"sheltered_at,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// IsFriends reports whether somebody else owns this plant.
func (p Plant) IsFriends() bool { return p.Steward != "" && p.Steward != StewardSelf }

// NeedsChasing reports whether a due action depends on a human remembering.
func (p Plant) NeedsChasing() bool { return p.WateringMethod == WateringHand }

// Risk ranks likelihood of neglect, not current thirst.
func (p Plant) Risk() int {
	risk := 0
	if p.NeedsChasing() {
		risk += 2
	}
	switch p.Accessibility {
	case AccessHard:
		risk += 2
	case AccessAwkward:
		risk++
	}
	if p.IsFriends() {
		risk += 2
	}
	if p.Status == StatusStruggling {
		risk++
	}
	return risk
}

// CareProfile is domain-shaped descriptive knowledge. Every field is optional,
// since most are unknown when a plant is first written down.
type CareProfile struct {
	Notes      string        `json:"notes,omitempty"`
	OwnerSays  string        `json:"owner_says,omitempty"`
	Quirks     string        `json:"quirks,omitempty"`
	WantsLight LightExposure `json:"wants_light,omitempty"`

	MoistureLow  *float64 `json:"moisture_low,omitempty"`
	MoistureHigh *float64 `json:"moisture_high,omitempty"`
	HumidityPref string   `json:"humidity_pref,omitempty"`
	Dormancy     string   `json:"dormancy,omitempty"`

	DaysToMaturity  *int       `json:"days_to_maturity,omitempty"`
	SownAt          *time.Time `json:"sown_at,omitempty"`
	TransplantedAt  *time.Time `json:"transplanted_at,omitempty"`
	FeedSchedule    string     `json:"feed_schedule,omitempty"`
	ExpectedHarvest *time.Time `json:"expected_harvest,omitempty"`

	// Indoors a tomato flowers and sets no fruit without something shaking the
	// flowers, and the plant looks healthy the whole time it yields nothing.
	NeedsPollination *bool `json:"needs_pollination,omitempty"`

	HardinessZone      string     `json:"hardiness_zone,omitempty"`
	FrostSensitive     *bool      `json:"frost_sensitive,omitempty"`
	SuccessionDays     *int       `json:"succession_days,omitempty"`
	LastFrostExpected  *time.Time `json:"last_frost_expected,omitempty"`
	FirstFrostExpected *time.Time `json:"first_frost_expected,omitempty"`
}

// Valid reports whether the enum fields hold values the store accepts, so a
// rejection reads the same whether the API or the Dusk plugin asked.
func (p Plant) Valid() error {
	switch p.Domain {
	case DomainHouseplant, DomainEdibleIndoor, DomainEdibleOutdoor:
	default:
		return invalid("unknown domain %q", p.Domain)
	}
	switch p.Status {
	case StatusAlive, StatusStruggling, StatusDormant, StatusDead, StatusGone:
	default:
		return invalid("unknown status %q", p.Status)
	}
	switch p.WateringMethod {
	case WateringLetPot, WateringHand:
	default:
		return invalid("unknown watering method %q", p.WateringMethod)
	}
	switch p.Accessibility {
	case AccessEasy, AccessAwkward, AccessHard:
	default:
		return invalid("unknown accessibility %q", p.Accessibility)
	}
	switch p.LightExposure {
	case "", LightDirect, LightBrightIndirect, LightMedium, LightLow:
	default:
		return invalid("unknown light exposure %q", p.LightExposure)
	}
	if p.WateringMethod == WateringHand && p.LetPotDripper != nil {
		return invalid("hand-watered plant cannot hold a LetPot dripper number")
	}
	if strings.TrimSpace(p.CommonName) == "" {
		return invalid("common_name is required")
	}
	return nil
}
